File: src/cc/redpen/intellij/SingleCharEditor.java
package cc.redpen.intellij;

import javax.swing.*;
import javax.swing.text.AttributeSet;
import javax.swing.text.BadLocationException;
import javax.swing.text.PlainDocument;

class SingleCharEditor extends DefaultCellEditor {
  public SingleCharEditor() {
    super(new JTextField(new SingleCharDocument(), null, 1));
  }

  @Override public boolean stopCellEditing() {
    return ((JTextField)getComponent()).getText().length() == 1 && super.stopCellEditing();
  }

  static class SingleCharDocument extends PlainDocument {
    public void insertString(int offset, String str, AttributeSet a) throws BadLocationException {
      if (str != null && str.length() + getLength() == 1) super.insertString(offset, str, a);
    }
  }
}

File: src/cc/redpen/intellij/SingleCharEditor.kt
package cc.redpen.intellij

import java.awt.Color
import javax.swing.DefaultCellEditor
import javax.swing.JComponent
import javax.swing.JTextField
import javax.swing.border.LineBorder
import javax.swing.text.AttributeSet
import javax.swing.text.PlainDocument

internal class SingleCharEditor : DefaultCellEditor(JTextField(SingleCharEditor.SingleCharDocument(), null, 1)) {
    init {
        (component as JComponent).border = LineBorder(Color.black)
    }

    override fun stopCellEditing(): Boolean {
        return (component as JTextField).text.length == 1 && super.stopCellEditing()
    }

    internal class SingleCharDocument : PlainDocument() {
        override fun insertString(offset: Int, str: String?, a: AttributeSet?) {
            if (str != null && str.length + length == 1) super.insertString(offset, str, a)
        }
    }
}
